taichi/backends/cuda/codegen_cuda.cpp

```diff
@@ -62,31 +62,55 @@ class CodeGenLLVMCUDA : public CodeGenLLVM {
-            // See CUDA driver API `cuPointerGetAttribute` for more details.
-            transferred = true;
-            CUDADriver::get_instance().malloc(&device_buffers[i], args[i].size);
-            CUDADriver::get_instance().memcpy_host_to_device(
-                (void *)device_buffers[i], arg_buffers[i], args[i].size);
-          } else {
-            device_buffers[i] = arg_buffers[i];
+          if (context.sizes_in_bytes[i] > 0) {
+            // Note: both numpy and PyTorch support arrays/tensors with zeros
+            // in shapes, e.g., shape=(0) or shape=(100, 0, 200). This makes
+            // args[i].size = 0.
+            unsigned int attr_val = 0;
+            uint32_t ret_code =
+                CUDADriver::get_instance().mem_get_attribute.call(
+                    &attr_val, CU_POINTER_ATTRIBUTE_MEMORY_TYPE,
+                    (void *)arg_buffers[i]);
+            if (ret_code != CUDA_SUCCESS || attr_val != CU_MEMORYTYPE_DEVICE) {
+              // Copy to device buffer if arg is on host
+              // - ret_code != CUDA_SUCCESS:
+              //   arg_buffers[i] is not on device
+              // - attr_val != CU_MEMORYTYPE_DEVICE:
+              //   Cuda driver is aware of arg_buffers[i] but it might be on
+              //   host.
+              // See CUDA driver API `cuPointerGetAttribute` for more details.
+              transferred = true;
+              CUDADriver::get_instance().malloc(&device_buffers[i],
+                                                args[i].size);
+              CUDADriver::get_instance().memcpy_host_to_device(
+                  (void *)device_buffers[i], arg_buffers[i], args[i].size);
+            } else {
+              device_buffers[i] = arg_buffers[i];
+            }
+            ctx_builder.set_arg_external_array(i, (uint64)device_buffers[i],
+                                               args[i].size,
+                                               /*is_device_allocation=*/false);
+
+          } else if (args[i].size > 0) {
+            // arg_buffers[i] is a DeviceAllocation*
+            // TODO: Unwraps DeviceAllocation* can be done at CodeGenLLVM since
+            // it's shared by cpu and cuda.
+            DeviceAllocation *ptr =
+                static_cast<DeviceAllocation *>(arg_buffers[i]);
+            device_buffers[i] = kernel->program->get_llvm_program_impl()
+                                    ->get_ndarray_alloc_info_ptr(*ptr);
+            // We compare arg_buffers[i] and device_buffers[i] later to check
+            // if transfer happened.
+            // TODO: this logic can be improved but I'll leave it to a followup
+            // PR.
+            arg_buffers[i] = device_buffers[i];
+
+            ctx_builder.set_arg_external_array(i, (uint64)device_buffers[i],
+                                               args[i].size,
+                                               /*is_device_allocation=*/false);
```

Comment:
`false` here is correct?

Reply:
Yup since we unwraps it and replaces it with a raw ptr, so it's no longer a device allocation.